bpo-36540: Documentation for PEP570 - Python positional only arguments (#13202)
python/cpython: Doc/tutorial/controlflow.rst

Note that the order in which the keyword arguments are printed is guaranteed
to match the order in which they were provided in the function call.

Special parameters
------------------

By default, arguments may be passed to a Python function either by position
or explicitly by keyword. For readability and performance, it makes sense to
restrict the way arguments can be passed so that a developer need only look
at the function definition to determine if items are passed by position, by
position or keyword, or by keyword.

A function definition may look like::

   def f(pos1, pos2, /, pos_or_kwd, *, kwd1, kwd2):

>          -----------    ----------     ----------

Reviewer: Distinct lack of [Box-drawing characters](https://en.wikipedia.org/wiki/Box-drawing_character), tut tut!

(For those who don't know that Pablo and I are friends, this is gentle ribbing, not me being a jerk 😉)

Author: Sphinx does something fancy with those characters and breaks the whole diagram :(

Author: @pganssle Do you know a code block in sphinx that does not break box-drawing characters?

Reviewer: I don't see anything fancy, am I missing something?

![box-chars](https://user-images.githubusercontent.com/1377457/57584240-d517a300-74a6-11e9-9419-80eb3366c516.png)

Looks like they work just fine (other than the fact that the vertical bars don't connect perfectly, but that's already true).

Author: The problem happens when you change all of the characters. Try it out and meanwhile I will try to add a screenshot.

Reviewer: ![box-chars](https://user-images.githubusercontent.com/1377457/57586793-0ef9a100-74c9-11e9-9a1d-5f5199e6ae59.png)

Still looks fine to me with all box characters.

Author: Can you copy here the text? I may be doing something wrong....

Reviewer: ```
.. code-block:: none

   def f(pos1, pos2, /, pos_or_kwd, *, kwd1, kwd2):
         ──┬───────     ─┬────────     ─────┬────
           │             │                  │
           │        Positional or keyword   │
           │                                └ Keyword only
           └── Positional only
```

This is what I have.

It renders weird for PDFs in older versions of Sphinx, but renders fine in both HTML and PDF with version 2.0.1.

Author: This is how it renders for me (copy-pasting you code):

![Screenshot from 2019-05-12 20-53-00](https://user-images.githubusercontent.com/11718525/57587140-fce32700-74f7-11e9-9f46-4066bc5e36b8.png)

Notice that in the editor it renders nicely:

![Screenshot from 2019-05-12 20-53-46](https://user-images.githubusercontent.com/11718525/57587143-16846e80-74f8-11e9-8e35-015021b81342.png)

Reviewer: Is this a problem with the font you are using in the browser? I tried this with two different browsers and it's rendering correctly.

Author: I changed the font a couple of times and I have the same problem :S

Author: On the other hand, firefox does it correctly. Can you try with Chrome?

Reviewer: @pablogsal It works fine on Chrome for me. Maybe we should leave it as is and open a new BPO issue about box characters not working correctly in documentation?

Seems like something that can probably be fixed with a webfont or something, but no need to hold up this PR on fixing it.

           |             |                  |
           |        Positional or keyword   |
           |                                - Keyword only
            -- Positional only

where ``/`` and ``*`` are optional. If used, these symbols indicate the kind of
parameter by how the arguments may be passed to the function:
positional-only, positional-or-keyword, and keyword-only. Keyword parameters
are also referred to as named parameters.

-------------------------------
Positional-or-Keyword Arguments
-------------------------------

If ``/`` and ``*`` are not present in the function definition, arguments may
be passed to a function by position or by keyword.

--------------------------
Positional-Only Parameters
--------------------------

Looking at this in a bit more detail, it is possible to mark certain parameters

> as *positional-only*. If *positional-only*, the parameters' order matters, and

Reviewer 2: Not sure "the parameters' order matters" adds anything. It is more than when calling by keyword order does not matter. on pos or keyword order can also matter if called via positional.

Author: This is for the case when we use the `range()` parameter order in the PEP. This part is also from the `How to teach this` from the PEP itself.

>It is more than when calling by keyword order does not matter. on pos or keyword order can also matter if called via positional.

I do not understand what you mean :S

Reviewer 2: I mean that if a function is declared with pos or keyword, the order matters in the same way as for postional only if called without keywords. Isn’t it?

Author: Yep, but here we are saying that if you use positional-only there is no way of calling with an altered order.

> the parameters cannot be passed by keyword. Positional-only parameters would

Reviewer: I think change "would be" to "are".

be placed before a ``/`` (forward-slash). The ``/`` is used to logically
separate the positional-only parameters from the rest of the parameters.
If there is no ``/`` in the function definition, there are no positional-only
parameters.

> Parameters following the ``/`` may be *positional-or-keyword* or *keyword-only*.

Reviewer 2: Parameters following the ``/`` are  *positional-or-keyword* unless specified otherwise by another marker.

Author: I prefer the original wording (it's from Carol's section in the PEP) :)

On the other hand, your sentence is a bit more concrete towards `positional-or-keyword`, but as there is only one other marker apart from "/", is not much more general to say "specified otherwise by another marker" IMHO

Reviewer 3: Good point @mariocj89 if the sentence was read by itself. I'm going to resolve this since the next section treats the other marker `*` and the code block before has a full example with `/` and `*`.

----------------------
Keyword-Only Arguments
----------------------

To mark parameters as *keyword-only*, indicating the parameters must be passed
by keyword argument, place an ``*`` in the arguments list just before the first
*keyword-only* parameter.

-----------------
Function Examples
-----------------

Consider the following example function definitions paying close attention to the
markers ``/`` and ``*``::

   >>> def standard_arg(arg):
   ...     print(arg)
   ...
   >>> def pos_only_arg(arg, /):
   ...     print(arg)
   ...
   >>> def kwd_only_arg(*, arg):
   ...     print(arg)
   ...
   >>> def combined_example(pos_only, /, standard, *, kwd_only):
   ...     print(pos_only, standard, kwd_only)

> The first function definition ``standard_arg``, the most familiar form,

Reviewer: Hm, I would put a comma before ``standard_arg`` here, but I think it's acceptable to leave the comma out. Up to you.

Same comment with the others.

   >>> combined_example(1, 2, kwd_only=3)
   1 2 3

   >>> combined_example(1, standard=2, kwd_only=3)
   1 2 3

   >>> combined_example(pos_only=1, standard=2, kwd_only=3)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   TypeError: combined_example() got an unexpected keyword argument 'pos_only'

Finally, consider this function definition::

    def foo(name, **kwds):
        return 'name' in kwds

There is no possible call that will make it return ``True`` as the keyword 'name'
will always to bind to the first parameter. For example::

    >>> foo(1, **{'name': 2})
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: foo() got multiple values for argument 'name'
    >>>

But using ``/`` (positional only arguments), this is possible::

    def foo(name, /, **kwds):
        return 'name' in kwds
    >>> foo(1, **{'name': 2})
    True

In other words, the names of positional-only parameters can be used in
``**kwds`` without ambiguity.

-----
Recap
-----

The use case will determine which parameters to use in the function definition::

   def f(pos1, pos2, /, pos_or_kwd, *, kwd1, kwd2):

As guidance:

> * Use positional-only if names do not matter or have no meaning, and there are

Reviewer 2: I  would remove the "do not matter" and try to clarify it a little further. Maybe something like:

Use positional-only if you want the name of the parameters to not be available to the user. This is useful when parameter names have no real meaning, if you want to enforce the order of the arguments when the function is called or if you need to take arbitrary keywords.

  only a few arguments which will always be passed in the same order or in case
  when you need arbitrary keyword parameters and some positional ones.

> * Use keyword-only when names have meaning and the function definition is

Reviewer 2: more understandable by being explicit with names or you want to prevent users relying on the position of the argument being passed.

   >>> pos_only_arg(1)
   1

   >>> pos_only_arg(arg=1)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   TypeError: pos_only_arg() got an unexpected keyword argument 'arg'

The third function ``kwd_only_args`` only allows keyword arguments as indicated
by a ``*`` in the function definition::

   >>> kwd_only_arg(3)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   TypeError: kwd_only_arg() takes 0 positional arguments but 1 was given

   >>> kwd_only_arg(arg=3)
   3

And the last uses all three calling conventions in the same function
definition::

   >>> combined_example(1, 2, 3)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   TypeError: combined_example() takes 2 positional arguments but 3 were given

   >>> combined_example(1, 2, kwd_only=3)
   1 2 3

   >>> combined_example(1, standard=2, kwd_only=3)
   1 2 3

   >>> combined_example(pos_only=1, standard=2, kwd_only=3)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   TypeError: combined_example() got an unexpected keyword argument 'pos_only'

> Finally, consider this function definition::

Reviewer 3: ```suggestion
Finally, consider this function definition which has a potential collision between the positional argument ``name``  and ``**kwds`` which has ``name`` as a key::
```

    def foo(name, **kwds):
        return 'name' in kwds

> There is no possible call that will make it return ``True`` as the keyword 'name'

Reviewer 3: ```suggestion
There is no possible call that will make it return ``True`` as the keyword ``'name'``
```

will always to bind to the first parameter. For example::

    >>> foo(1, **{'name': 2})
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: foo() got multiple values for argument 'name'
    >>>

> But using ``/`` (positional only arguments), this is possible::

Reviewer 3: ```suggestion
But using ``/`` (positional only arguments), it is possible since it allows ``name`` as a positional argument and ``'name'`` as a key in the keyword arguments.
```

> But using ``/`` (positional only arguments), it is possible since it allows ``name`` as a positional argument and ``'name'`` as a key in the keyword arguments.

Reviewer 3: ```suggestion
But using ``/`` (positional only arguments), it is possible since it allows ``name`` as a positional argument and ``'name'`` as a key in the keyword arguments::
```

Reviewer 3: @pablogsal I think the suggestion will correct the build error.

Author: Thanks, @willingc  I was accepting the suggestions from the phone (best present ever) and could not investigate manually :)

    def foo(name, /, **kwds):
        return 'name' in kwds
    >>> foo(1, **{'name': 2})
    True

In other words, the names of positional-only parameters can be used in
``**kwds`` without ambiguity.

-----
Recap
-----

The use case will determine which parameters to use in the function definition::

   def f(pos1, pos2, /, pos_or_kwd, *, kwd1, kwd2):

As guidance:

* Use positional-only if you want the name of the parameters to not be
  available to the user. This is useful when parameter names have no real

>   meaning, if you want to enforce the order of the arguments when the function

Reviewer 4: `Or if you believe the name of the parameter may change in the future, and a change of parameter name can be considered a change in API` ? 

I've seen a couple of time `def foo(filename)` -> `def foo(filename_or_buffer)`, which people don't realise is an api change.

Reviewer 3: Perhaps at end of these bullets, add a bullet. For an API, use positional only to prevent breaking API changes if the parameter's name changes.

>   is called or if you need to take some positinal parameters and arbitrary

Reviewer 3: ```suggestion
  is called or if you need to take some positional parameters and arbitrary
```
